src/aiortc/contrib/media.py

```diff
@@ -171,7 +172,7 @@ def __init__(self, player, kind):
         frame_time = frame.time
 
         # control playback rate
-        if self._player._throttle_playback and frame_time is not None:
```

Comment:
Deadlock was happening here. At the time the first expression in the condition is being checked by a thread, another thread is closing the player, due to its tracks closure and it would become `None`.

Reply:
I'm not sure I understand this, there should be no threading involved here: everything happens on the main thread, except for `player_worker` which does not touch the `Player` or `PlayerStreamTrack` objects.

Reply:
@jlaine,

I'm not sure  I understand it neither. Flow is blocked in such line, just put a log before and after it. 

git checkout 2b4fce4b58aa3c56a05a112ab4d0634e0c45a3a6 in the PR and run the new tests added in such commit (those retrieving the system webcam/mic)

Hope it helps.

Reply:
OK, I don't think this is actually a threading problem, but you are right: flow can block a few lines higher on the `frame = await self._queue.get()` line since it's an async statement, so `self._player` may have been cleared in the meantime. This would result in an exception in the caller (`RTCRtpSender`) which would therefore never properly reach the point where it sets the `_rtp_exited` event.

Reply:
I think two changes would make sense here:

- check `self._player is not None` as you did
- make `RTCRtpSender._run_rtp` more robust, by moving its shutdown code (including the critical `self.__rtp_exited.set()` in a `finally` clause of the try / catch

Reply:
@jmillan in a5764dc163351394f61108cd39f7351608fd90bc I made a change which ensures `RTCRtpSender._run_rtp` now catches and logs any unexpected exceptions and exits cleanly. Could you try running your tests without any of your fixes, you should now see an `AttributeError`?

Reply:
I'm not running those tests but indeed, after `git pull` in aiortc "master", I see such an error:

```
sender(video) - RTP worker encountered an exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/aiortc/rtcrtpsender.py", line 295, in _run_rtp
    payloads, timestamp = await self._next_encoded_frame(codec)
  File "/usr/local/lib/python3.7/site-packages/aiortc/rtcrtpsender.py", line 248, in _next_encoded_frame
    frame = await self.__track.recv()
  File "/usr/local/lib/python3.7/site-packages/aiortc/contrib/media.py", line 174, in recv
    if self._player._throttle_playback and frame_time is not None:
AttributeError: 'NoneType' object has no attribute '_throttle_playback'
```

Reply:
I've done some tests and I may say that, after that commit, I am not able to reproduce the "zombie process" issue anymore. This is, no matter I stop the mic/webcam tracks before/after closing the `pcs`, there are no more Python zombie processes. This is good.

Reply:
OK that is excellent news. I've had to back the commit out of master as CI was failing and I've moved the code to PR #293 to examine what's going on. Sorry for the merge conflict you're about to experience..

Reply:
n.p. BTW I'm testing `transceiver.sender.replaceTrack()` and I get an error:

```
sender(video) - RTP worker encountered an exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/aiortc/rtcrtpsender.py", line 295, in _run_rtp
    payloads, timestamp = await self._next_encoded_frame(codec)
  File "/usr/local/lib/python3.7/site-packages/aiortc/rtcrtpsender.py", line 248, in _next_encoded_frame
    frame = await self.__track.recv()
  File "/usr/local/lib/python3.7/site-packages/aiortc/contrib/media.py", line 174, in recv
    if self._player._throttle_playback and frame_time is not None:
AttributeError: 'NoneType' object has no attribute '_throttle_playback'
```

Reply:
This is very much related, it's the same problem @jmillan pointed to, and which is addressed by PR #294. The difference is that we now have a visible error rather than a silent deadlock.

Reply:
Sorry if unrelated. Randomly, `replaceTrack()` does not fail (no error log), however no RTP is sent anymore in the corresponding transceiver. This is probably related to the stuff you are already fixing :)

Reply:
Well, sorry for the noise. Actually, when `replaceTrack()` works (and I've remove the `track.stop()` in `rtcrtpsender.py`), it keeps sending RTP. Confirmed. However the remote (Chrome) cannot render those RTP during ~10 seconds and gets frozen video. Asking for a key frame via RTCP PLI does not make it work. After ~10 seconds, the receiver can render the video again.